src/graph/optimizer/rule/PushFilterDownHashInnerJoinRule.cpp

```diff
@@ -0,0 +1,136 @@
+  OptGroup* leftGroup = pushFilterDownChild(octx, leftResult, condition, &leftFilterUnpicked);
+  OptGroup* rightGroup =
+      pushFilterDownChild(octx, rightResult, leftFilterUnpicked, &rightFilterUnpicked);
+
+  if (!leftGroup && !rightGroup) {
+    return TransformResult::noTransform();
+  }
+
+  leftGroup = leftGroup ? leftGroup : const_cast<OptGroup*>(leftResult.node->group());
+  rightGroup = rightGroup ? rightGroup : const_cast<OptGroup*>(rightResult.node->group());
+
+  // produce new InnerJoin node
+  auto* newInnerJoinNode = static_cast<graph::HashInnerJoin*>(oldInnerJoinNode->clone());
+  auto newJoinGroup = rightFilterUnpicked ? OptGroup::create(octx) : filterGroupNode->group();
+  // TODO(yee): it's too tricky
+  auto newGroupNode = rightFilterUnpicked
+                          ? const_cast<OptGroup*>(newJoinGroup)->makeGroupNode(newInnerJoinNode)
+                          : OptGroupNode::create(octx, newInnerJoinNode, newJoinGroup);
+  newGroupNode->dependsOn(leftGroup);
+  newGroupNode->dependsOn(rightGroup);
+  newInnerJoinNode->setLeftVar(leftGroup->outputVar());
+  newInnerJoinNode->setRightVar(rightGroup->outputVar());
+
+  if (rightFilterUnpicked) {
+    auto newFilterNode = graph::Filter::make(octx->qctx(), nullptr, rightFilterUnpicked);
+    newFilterNode->setOutputVar(oldFilterNode->outputVar());
+    newFilterNode->setColNames(oldFilterNode->colNames());
+    newFilterNode->setInputVar(newInnerJoinNode->outputVar());
+    newGroupNode = OptGroupNode::create(octx, newFilterNode, filterGroupNode->group());
+    newGroupNode->dependsOn(const_cast<OptGroup*>(newJoinGroup));
+  } else {
+    newInnerJoinNode->setOutputVar(oldFilterNode->outputVar());
+    newInnerJoinNode->setColNames(oldFilterNode->colNames());
+  }
+
+  TransformResult result;
+  result.eraseAll = true;
+  result.newGroupNodes.emplace_back(newGroupNode);
+  return result;
+}
+
+OptGroup* PushFilterDownHashInnerJoinRule::pushFilterDownChild(OptContext* octx,
+                                                               const MatchedResult& child,
+                                                               const Expression* condition,
+                                                               Expression** unpickedFilter) {
+  if (!condition) return nullptr;
+
+  const auto* childPlanNode = DCHECK_NOTNULL(child.node->node());
+  const auto& colNames = childPlanNode->colNames();
+
+  // split the `condition` based on whether the varPropExpr comes from the left child
+  auto picker = [&colNames](const Expression* e) -> bool {
+    return graph::ExpressionUtils::checkVarPropIfExist(colNames, e);
+  };
+
+  Expression* filterPicked = nullptr;
+  graph::ExpressionUtils::splitFilter(condition, picker, &filterPicked, unpickedFilter);
+  if (!filterPicked) return nullptr;
+
+  auto* newChildPlanNode = childPlanNode->clone();
+  DCHECK_NE(childPlanNode->outputVar(), newChildPlanNode->outputVar());
```

Comment:
Just the behavior of PlanNode::PlanNode(), need DCHECK here?

Reply:
Yeah, This line could be removed!